### PR TITLE
Refactor capability checks for ACT's

### LINKED ISSUE #5503: Roles & Capabilities pods_edit_others not working with advanced content type
**Describe the bug**

I created an advanced content type named "tcat1".
In "Roles & Capabilities" for Editor role :
I checked pods_add_tcat1, pods_delete_tcat1 and pods_edit_tcat1
I unchecked pods_edit_others_tcat1 and pods_delete_others_tcat1.

I created 2 Editor users, each of them created a record and they can modify the record of the other.

**To Reproduce**

**Expected behavior**
A user should not be able to modify the record of another user if pods_edit_others_ is not checked for his role.

**Screenshots**

## Pods Version
Pods Version 2.7.15

## WordPress Environment

<details>

Copy and paste your Tools > Site Health > Info. Use the 'Copy Site Info to Clipboard' and paste the results over this line.
`
### wp-core ###

version: 5.2.4
site_language: en_US
user_language: en_US
permalink: undefined
https_status: true
user_registration: 0
default_comment_status: open
multisite: false
user_count: 3
dotorg_communication: true

### wp-paths-sizes ###

wordpress_path: /home/cmcweb/swww/tests/tw1
wordpress_size: 40.03 MB (41976209 bytes)
uploads_path: /home/cmcweb/swww/tests/tw1/wp-content/uploads
uploads_size: 0.00 B (0 bytes)
themes_path: /home/cmcweb/swww/tests/tw1/wp-content/themes
themes_size: 3.06 MB (3208811 bytes)
plugins_path: /home/cmcweb/swww/tests/tw1/wp-content/plugins
plugins_size: 5.35 MB (5610537 bytes)
database_size: 2.92 MB (3063808 bytes)
total_size: 51.36 MB (53859365 bytes)

### wp-active-theme ###

name: Twenty Nineteen
version: 1.4
author: the WordPress team
author_website: https://wordpress.org/
parent_theme: none
theme_features: automatic-feed-links, title-tag, post-thumbnails, menus, html5, custom-logo, customize-selective-refresh-widgets, wp-block-styles, align-wide, editor-styles, editor-style, editor-font-sizes, editor-color-palette, responsive-embeds, widgets
theme_path: /home/cmcweb/swww/tests/tw1/wp-content/themes/twentynineteen

### wp-themes (2) ###

Twenty Seventeen: version: 2.2, author: the WordPress team
Twenty Sixteen: version: 2.0, author: the WordPress team

### wp-plugins-active (1) ###

Pods - Custom Content Types and Fields: version: 2.7.15, author: Pods Framework Team

### wp-plugins-inactive (2) ###

Akismet Anti-Spam: version: 4.1.2, author: Automattic
Hello Dolly: version: 1.7.2, author: Matt Mullenweg

### wp-media ###

image_editor: WP_Image_Editor_GD
imagick_module_version: Not available
imagemagick_version: Not available
gd_version: 2.2.5
ghostscript_version: not available

### wp-server ###

server_architecture: Linux 4.19.0-5-cloud-amd64 x86_64
httpd_software: nginx/1.14.2
php_version: 7.3.9-1~deb10u1 64bit
php_sapi: fpm-fcgi
max_input_variables: 1000
time_limit: 30
memory_limit: 256M
max_input_time: 60
upload_max_size: 2M
php_post_max_size: 8M
curl_version: 7.64.0 OpenSSL/1.1.1d
suhosin: false
imagick_availability: false

### wp-database ###

extension: mysqli
server_version: 5.5.5-10.3.17-MariaDB-0+deb10u1
client_version: mysqlnd 5.0.12-dev - 20150407 - $Id: 7cc7cc96e675f6d72e5cf0f267f48e167c2abb23 $

### wp-constants ###

WP_HOME: undefined
WP_SITEURL: undefined
WP_CONTENT_DIR: /home/cmcweb/swww/tests/tw1/wp-content
WP_PLUGIN_DIR: /home/cmcweb/swww/tests/tw1/wp-content/plugins
WP_MAX_MEMORY_LIMIT: 256M
WP_DEBUG: false
WP_DEBUG_DISPLAY: true
WP_DEBUG_LOG: false
SCRIPT_DEBUG: false
WP_CACHE: false
CONCATENATE_SCRIPTS: undefined
COMPRESS_SCRIPTS: undefined
COMPRESS_CSS: undefined
WP_LOCAL_DEV: undefined

### wp-filesystem ###

wordpress: writable
wp-content: writable
uploads: writable
plugins: writable
themes: writable

### pods ###

pods-server-software: nginx/1.14.2
pods-user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0
pods-session-save-path: /var/lib/php/sessions
pods-session-save-path-exists: Yes
pods-session-save-path-writable: Yes
pods-session-max-lifetime: 1440
pods-opcode-cache-apc: No
pods-opcode-cache-memcached: No
pods-opcode-cache-opcache: Yes
pods-opcode-cache-redis: No
pods-object-cache-apc: No
pods-object-cache-apcu: No
pods-object-cache-memcache: No
pods-object-cache-memcached: No
pods-object-cache-redis: No
pods-memory-current-usage: 4.299M
pods-memory-current-usage-real: 4.000M
pods-network-wide: No
pods-install-location: /home/cmcweb/swww/tests/tw1/wp-content/plugins/pods/
pods-developer: No
pods-tableless-mode: No
pods-light-mode: No
pods-strict: No
pods-allow-deprecated: Yes
pods-api-cache: Yes

`

</details>

## Pods Package Export (helpful!)

```
{"meta":{"version":"2.7.15","build":1571402834},"pods":{"27":{"id":27,"name":"tcat1","label":"tcat1","description":"","type":"pod","storage":"table","object":"","alias":"","fields":{"adresse_mac":{"id":33,"name":"adresse_mac","label":"Adresse MAC","description":"Texte decrivant l'<a href=\"\">adresse<\/a> MAC","help":"","class":"","type":"text","weight":0,"pick_object":"","pick_val":"","sister_id":"","required":"0","unique":"0","text_allow_shortcode":"0","text_allow_html":"0","text_allowed_html_tags":"strong em a ul ol li b i","text_max_length":"32","pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"text_repeatable":"0","text_placeholder":""},"adresse_ip":{"id":34,"name":"adresse_ip","label":"Adresse IP","description":"","help":"","class":"","type":"text","weight":1,"pick_object":"","pick_val":"","sister_id":"","required":"0","unique":"0","text_allow_shortcode":"0","text_allow_html":"0","text_allowed_html_tags":"strong em a ul ol li b i","text_max_length":"32","pick_post_status":["publish"],"admin_only":"0","restrict_role":"0","restrict_capability":"0","hidden":"0","read_only":"0","roles_allowed":["administrator"],"text_repeatable":"0","text_placeholder":""},"created":{"id":29,"name":"created","label":"Date Created","description":"","help":"","class":"","type":"datetime","weight":2,"pick_object":"","pick_val":"","sister_id":"","datetime_format":"ymd_slash","datetime_time_type":"12","datetime_time_format":"h_mm_ss_A","required":"0","unique":"0","datetime_repeatable":"0","datetime_type":"format","datetime_time_format_24":"hh_mm","datetime_allow_empty":"1","datetime_html5":"0","datetime_format_custom":"","datetime_format_custom_js":"","datetime_time_format_custom":"","datetime_time_format_custom_js":"","datetime_year_range_custom":""},"modified":{"id":30,"name":"modified","label":"Date Modified","description":"","help":"","class":"","type":"datetime","weight":3,"pick_object":"","pick_val":"","sister_id":"","datetime_format":"ymd_slash","datetime_time_type":"12","datetime_time_format":"h_mm_ss_A","required":"0","unique":"0","datetime_repeatable":"0","datetime_type":"format","datetime_time_format_24":"hh_mm","datetime_allow_empty":"1","datetime_html5":"0","datetime_format_custom":"","datetime_format_custom_js":"","datetime_time_format_custom":"","datetime_time_format_custom_js":"","datetime_year_range_custom":""},"author":{"id":31,"name":"author","label":"Author","description":"","help":"","class":"","type":"pick","weight":4,"pick_object":"user","pick_val":"","sister_id":"","pick_format_type":"single","pick_format_single":"autocomplete","default_value":"{@user.ID}","required":"0","unique":"0","pick_format_multi":"checkbox","pick_allow_add_new":"1","pick_taggable":"0","pick_show_icon":"1","pick_show_edit_link":"1","pick_show_view_link":"1","pick_limit":"0","pick_post_status":"publish","pick_select_text":"","pick_table_id":"","pick_table_index":"","pick_display":"","pick_user_role":"","pick_where":"","pick_orderby":"","pick_groupby":""},"permalink":{"id":32,"name":"permalink","label":"Permalink","description":"Leave blank to auto-generate from Name","help":"","class":"","type":"slug","weight":5,"pick_object":"","pick_val":"","sister_id":"","required":"0","unique":"0","slug_placeholder":""}},"show_in_menu":"1","label_singular":"tcat1","pod_index":"adresse_mac","hierarchical":"0","ui_style":"post_type","menu_position":"0","ui_actions_enabled":["add","edit","duplicate","delete"],"ui_reorder_field":"menu_order","ui_fields_manage":["adresse_mac","adresse_ip","created","modified","author"],"ui_filters":{"0":"adresse_mac","1":"adresse_ip","4":"author"}}}}

```

**Additional context**

**Possible Workaround**
If you have discovered a workaround, please include it below.

### PR DESCRIPTION
**Pull request by JoryHogeveen**
Fixes #5503
Closes #5176 

## Description

As stated in the issues above the capabilities for ACT's do not work properly. This PR fixes the access checks for ACT's.

## ChangeLog

* Refactor and fix issues with Advanced Content Types capabilities.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: classes/PodsUI.php [hunk 1]
```diff
@@ -4927,16 +4927,43 @@ public function restricted( $action = 'edit', $row = null ) {
...
-					if ( ! in_array( 'edit', $this->actions_disabled ) && ( current_user_can( 'pods_edit_' . $this->pod->pod ) || current_user_can( 'pods_edit_others_' . $this->pod->pod ) ) ) {
-						$restricted = false;
-					} elseif ( ! in_array( 'delete', $this->actions_disabled ) && ( current_user_can( 'pods_delete_' . $this->pod->pod ) || current_user_can( 'pods_delete_others_' . $this->pod->pod ) ) ) {
-						$restricted = false;
-					} elseif ( current_user_can( 'pods_' . $action . '_' . $this->pod->pod ) || current_user_can( 'pods_' . $action . '_others_' . $this->pod->pod ) ) {
-						$restricted = false;
+				} else {
+
+					// Check if the current user is the author of this item.
+					$author = pods( $this->pod->pod, $row )->field( 'author', true );
```
Comment (tr1b0t) [inline classes/PodsUI.php]: :warning: Equals sign not aligned with surrounding assignments; expected 4 spaces but found 1 space

<sub>hash: ddfd2bf3441648bbe8c5ca120f39821f</sub>

File: classes/PodsUI.php [hunk 2]
```diff
@@ -4927,16 +4927,43 @@ public function restricted( $action = 'edit', $row = null ) {
...
+					$is_author = false;
+					if ( $author ) {
+						$is_author = (int) $author->ID === (int) wp_get_current_user()->ID;
```
Comment (tr1b0t) [inline classes/PodsUI.php]: * :no_entry_sign: The value of a comparison must not be assigned to a variable
* :no_entry_sign: Use Yoda Condition checks, you must.

<sub>hash: 7a8145e4f311c3735446b9b49bfef220</sub>

File: classes/PodsUI.php [hunk 3]
```diff
@@ -4927,16 +4927,43 @@ public function restricted( $action = 'edit', $row = null ) {
...
+					}
+
+					$cap_actions = array( $action );
+					if ( 'manage' === $action || 'reorder' === $action ) {
+						if ( ! in_array( 'edit', $this->actions_disabled ) ) {
```
Comment (tr1b0t) [inline classes/PodsUI.php]: :warning: Not using strict comparison for in_array; supply true for third argument.

<sub>hash: 2d8ec269a7ae11eaee044ece22ee324e</sub>

File: classes/PodsUI.php [hunk 4]
```diff
@@ -4927,16 +4927,43 @@ public function restricted( $action = 'edit', $row = null ) {
...
+							$cap_actions[] = 'edit';
+						}
+						if ( ! in_array( 'delete', $this->actions_disabled ) ) {
```
Comment (tr1b0t) [inline classes/PodsUI.php]: :warning: Not using strict comparison for in_array; supply true for third argument.

<sub>hash: f415106856953401b52327595e9f306b</sub>

**Comment by sdadoun** [comment]
Hello, 
  thank you for this quick Pull request, I am not very familiar with github, but I think I tried your modification ($git clone pods.git and $git pr 5504) and it properly displays only the pods's user and allow the user to delete it but when the user tries to edit it, it displays the error : 
Error: You do not have access to this tcat1.
(tcat1 is the name I gave to the ACT).
What is the proper way to report this problem ?

**Comment by JoryHogeveen** [comment]
Hello @sdadoun 
Could you verify that your user has the correct capabilities?

**Comment by sdadoun** [comment]
Hello Jory, thank for your quick reply.
Yes my user has Editor role which has delete and edit capabilities for the ACT I created (I double check by installing User Role Editor plugin).
I also created a new ACT tcat2 to be sure, when no tcat2 capabilities is checked nothing appear ; with pods_add_tcat2 checked I only see the menu tcat2 and I can add a tcat2 pods ; with either pods_delete_tcat2 and/or pods_edit_tcat2 checked I see "All tcat2" sub-menu and I can always delete my pods but never edit them.

**Comment by JoryHogeveen** [comment]
Hi @sdadoun 
I managed to reproduce your issue and fixed it in my latest commit. Could you also test this?

**Comment by sdadoun** [comment]
Hi Jorry, I just tested the latest commit and almost everything seems to work as expected.
User with edit and delete capabilities can only edit and delete their own pods.
User with edit_others and delete_others capabilities can edit and delete other pods.
But user with edit but not delete capability seems still able to delete its own pods.
For my use case it is enough, but I can continue to test if you want.
Thanks a lot.
Regards

File: classes/PodsUI.php [hunk 5]
```diff
@@ -4927,16 +4927,48 @@ public function restricted( $action = 'edit', $row = null ) {
...
-						$restricted = false;
-					} elseif ( ! in_array( 'delete', $this->actions_disabled ) && ( current_user_can( 'pods_delete_' . $this->pod->pod ) || current_user_can( 'pods_delete_others_' . $this->pod->pod ) ) ) {
-						$restricted = false;
-					} elseif ( current_user_can( 'pods_' . $action . '_' . $this->pod->pod ) || current_user_can( 'pods_' . $action . '_others_' . $this->pod->pod ) ) {
-						$restricted = false;
+				} else {
+
+					$pod = $this->pod;
+					if ( ! $pod->id() && $row ) {
+						$pod = $pod->fetch( $row );
```
Comment (sc0ttkclark) [inline classes/PodsUI.php]:
```suggestion
						$pod->fetch( $row );
```

**Comment by sc0ttkclark** [comment]
Let's see what we can do about getting capability check working properly here for delete.

**Comment by JoryHogeveen** [comment]
@sc0ttkclark  Capability checks for delete fixed.
The restriction code of PodsUI will need a review when it's up for refactor.

File: classes/PodsUI.php [hunk 6]
```diff
@@ -5088,50 +5112,38 @@ public function restricted( $action = 'edit', $row = null ) {
...
-										$okay = true;
-									}
+					if ( ! empty( $author_restricted ) ) {
+						if ( ! is_array( $author_restricted ) ) {
+							$author_restricted = (array) $author_restricted;
+						}
+
+						if ( is_array( $val ) ) {
+							foreach ( $val as $v ) {
+								if ( in_array( $v, $author_restricted ) ) {
```
Comment (tr1b0t) [inline classes/PodsUI.php]: :warning: Not using strict comparison for in_array; supply true for third argument.

<sub>hash: 59f60b68f4e03c4c915286329994a08a</sub>

File: classes/PodsUI.php [hunk 7]
```diff
@@ -5088,50 +5112,38 @@ public function restricted( $action = 'edit', $row = null ) {
...
+									$restricted = false;
 								}
-							} elseif ( in_array( $val, $author_restricted ) ) {
-								$okay = true;
 							}
+						} elseif ( in_array( $val, $author_restricted ) ) {
```
Comment (tr1b0t) [inline classes/PodsUI.php]: :warning: Not using strict comparison for in_array; supply true for third argument.

<sub>hash: 915ecd731834ac67611b18245fc4baf3</sub>

**Pull request merged**